### PR TITLE
Refactor: Improve string quoting logic and fixed string-related bugs

### PR DESCRIPTION
**Pull request by liplum**
This pull request bumps the package to 2.1.0.
It doesn't have breaking changes on API level, but the end-user's output result might be changed along with the logic changes.

### CODE REVIEW

File: lib/src/node.dart [hunk 1]
```diff
@@ -56,30 +60,43 @@ class StringNode extends Node {
...
-      r'[^0-9a-zA-ZàèìòùÀÈÌÒÙáéíóúýÁÉÍÓÚÝâêîôûÂÊÎÔÛãñõÃÑÕäëïöüÿÄËÏÖÜŸçÇßØøÅåÆæœ@/. \t-]');
+  static final _invalidCharsRegex = RegExp(
+    r'^[:{}\[\]>,&*#?|@-]|^\s+|\s+$|\n|\t|^[0-9]+$|^[0-9]*\.[0-9]+$',
+  );
 
-  bool _isValidUnquotedString(String s) =>
-      !_regexpInvalidUnquotedChars.hasMatch(s) &&
-      !s.startsWith('@') &&
-      !s.startsWith('-');
+  bool _isValidUnquotedString(String s) => !_invalidCharsRegex.hasMatch(s);
```
Comment (gmpassos) [inline lib/src/node.dart]: Could you please add a unit test for this method? You may expose it as public and static.

File: lib/src/yaml_writer.dart [hunk 2]
```diff
@@ -12,25 +12,15 @@ typedef YAMLWriter = YamlWriter;
...
-  YamlWriter({
-    int indentSize = 2,
-    this.allowUnquotedStrings = false,
-    Object? Function(dynamic object)? toEncodable,
-  })  : indentSize = max(1, indentSize),
-        toEncodable = toEncodable ?? _defaultToEncodable;
+  const YamlWriter({
+    this.config = const YamlWriterConfig(),
+    this.toEncodable = _defaultToEncodable,
+  });
```
Comment (gmpassos) [inline lib/src/yaml_writer.dart:43]: How about adding a new factory constructor, `YamlWriter.config`, for these new parameters?  

This way, we can keep the unnamed constructor with the original parameters to avoid breaking the API.

**Review by gmpassos — COMMENTED** [review]
Thanks for the PR.

File: lib/src/node.dart [hunk 3]
```diff
@@ -56,30 +67,42 @@ class StringNode extends Node {
...
       }
     }
 
     return yamlLines;
   }
 
-  static final _regexpInvalidUnquotedChars = RegExp(
-      r'[^0-9a-zA-ZàèìòùÀÈÌÒÙáéíóúýÁÉÍÓÚÝâêîôûÂÊÎÔÛãñõÃÑÕäëïöüÿÄËÏÖÜŸçÇßØøÅåÆæœ@/. \t-]');
+  static final _invalidCharsRegex = RegExp(
+    r'^[:{}\[\]>,&*#?|@-]|^\s+|\s+$|\n|\t|^[0-9]+$|^[0-9]*\.[0-9]+$',
```
Comment (gmpassos) [inline lib/src/node.dart]: add `\r`

File: lib/src/yaml_writer.dart (hunk 2 above)
Comment (liplum) [reply lib/src/yaml_writer.dart:43]: Done. And I bump the package version back to 2.2.0.

File: lib/src/node.dart (hunk 3 above)
Comment (liplum) [reply lib/src/node.dart]: Done.

File: lib/src/node.dart (hunk 1 above)
Comment (liplum) [reply lib/src/node.dart]: Done. I made a bunch of tests including edge cases.

**Comment by liplum** [comment]
I've added more test cases, and all tests are passed.

File: README.md [hunk 4]
```diff
@@ -75,9 +75,10 @@ Please file feature requests and bugs at the [issue tracker][tracker].
 
 ## Author
 
-Graciliano M. Passos: [gmpassos@GitHub][github].
+Graciliano M. Passos: [gmpassos@GitHub](https://github.com/gmpassos).
+
+Liplum: [Github](https://github.com/liplum)
```
Comment (gmpassos) [inline README.md]: liplum@GitHub

File: test/quote_test.dart [hunk 5]
```diff
@@ -0,0 +1,135 @@
+import 'package:test/test.dart';
+import 'package:yaml_writer/src/node.dart';
+
+void main() {
+  group('Unquoted', () {
+    test("simple text", () {
+      expect(StringNode.isValidUnquotedString("foo"), equals(true));
```
Comment (gmpassos) [inline test/quote_test.dart]: change `equals(true)` to `isTrue`

File: test/quote_test.dart [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      expect(StringNode.isValidUnquotedString("FOO"), equals(true));
+      expect(StringNode.isValidUnquotedString("foo-bar"), equals(true));
+      expect(StringNode.isValidUnquotedString("foo bar"), equals(true));
+    });
+
+    test("numbers", () {
+      // numbers are invalid because yaml would consider them as numbers other than strings
+      expect(StringNode.isValidUnquotedString("1"), equals(false));
```
Comment (gmpassos) [inline test/quote_test.dart]: Change `equals(false)` to `isFalse`

**Comment by gmpassos** [comment]
I think that after these small adjustments, it's okay to merge.

File: README.md (hunk 4 above)
Comment (liplum) [reply README.md]: Done

File: test/quote_test.dart (hunk 6 above)
Comment (liplum) [reply test/quote_test.dart]: Done

File: test/quote_test.dart (hunk 5 above)
Comment (liplum) [reply test/quote_test.dart]: Done

**Comment by liplum** [comment]
I think it's ready to merge.

**Comment by liplum** [comment]
I think most edge cases are tested.

File: lib/src/node.dart [hunk 7]
```diff
@@ -56,30 +67,100 @@ class StringNode extends Node {
...
-      } else if (!containsSingleQuote) {
-        yamlLines.add('\'$text\'');
+      final containsSingleQuote = text.contains("'");
+      final containsDoubleQuote = text.contains('"');
+
+      if (containsSingleQuote && containsDoubleQuote) {
+        // if contains both single and double quote, escape the double quote,
+        // and use double quote
+        final result = text.replaceAll(r'\', r'\\').replaceAll('"', r'\"');
+        yamlLines.add('"$result"');
```
Comment (gmpassos) [inline lib/src/node.dart]: Use `context.config.quoteStyle` to decide which quote to use for this string case.

Comment (liplum) [reply lib/src/node.dart]: I don't think single quote can be used for this case. Could you give me an example that single quote works in both single and double quote exisiting suitation?

**Comment by liplum** [comment]
Now I use `QuoteStyle` to decide the empty string output instead of `EmptyStringLiteral` to reduce complexity.

File: lib/src/node.dart (hunk 7 above)
Comment (gmpassos) [reply lib/src/node.dart]: Well, this is a single-line string, so it won't escape `\r`, `\n`, or `\t`.  

It only needs to escape:  

- Double quotes (`"`): `\"` (when using `preferDoubleQuote`)  
- Single quotes (`'`): `''` (when using `preferSingleQuote`; two single quotes escape it)  

This ensures that the `QuoteStyle` is maintained.

Comment (gmpassos) [reply lib/src/node.dart]: Do you think that we need to maintain the `QuoteStyle` in this case?

Comment (liplum) [reply lib/src/node.dart]: Done

File: lib/src/node.dart [hunk 8]
```diff
@@ -56,30 +68,89 @@ class StringNode extends Node {
...
-        yamlLines.add('"$str"');
+        var result = text;
+        switch (context.config.quoteStyle) {
+          case QuoteStyle.singleQuote:
+            result = result.replaceAll("'", "''");
+            yamlLines.add("'$result'");
+            break;
+          case QuoteStyle.doubleQuote:
+            result = result.replaceAll('"', r'\"');
+            yamlLines.add('"$result"');
```
Comment (gmpassos) [inline lib/src/node.dart:84]: With double quotes, you still need to escape any `\` character.

Comment (gmpassos) [reply lib/src/node.dart:84]: Please, add a unit test for escaped characters, and if it can be read by
https://pub.dev/packages/yaml
(added as dev_dependency, only for tests).

Comment (liplum) [reply lib/src/node.dart:84]: > With double quotes, you still need to escape any `\` character.

Now I escape the `\` char.

Comment (liplum) [reply lib/src/node.dart:84]: > Please, add a unit test for escaped characters, and if it can be read by https://pub.dev/packages/yaml (added as dev_dependency, only for tests).

Tests are added.

**Comment by gmpassos** [comment]
Merged.  

Thanks for your great work! I will publish it now.
